File: core/recc/packet/daemon.py
# -*- coding: utf-8 -*-

from dataclasses import dataclass
from datetime import datetime
from enum import Enum, unique
from typing import Optional


@dataclass
class Daemon:
    """It is mapped to the `daemon` table in the database."""

    uid: int
    plugin: str
    slug: str
    name: str
    address: str
    description: str
    enable: bool
    created_at: datetime
    updated_at: datetime


@unique
class DaemonState(Enum):
    Unknown = 0
    Unregistered = 1
    Down = 2
    NoSuchProcess = 3

    Running = 101
    Sleeping = 102
    DiskSleep = 103
    Stopped = 104
    TracingStop = 105
    Zombie = 106
    Dead = 107
    WakeKill = 108
    Waking = 109
    Idle = 110
    Locked = 111
    Waiting = 112
    Suspended = 113
    Parked = 114

    @property
    def invalid(self) -> bool:
        return self.value < 100

    @classmethod
    def from_process_status(cls, status: str):
        if status == "running":
            return cls.Running
        elif status == "sleeping":
            return cls.Sleeping
        elif status == "disk-sleep":
            return cls.DiskSleep
        elif status == "stopped":
            return cls.Stopped
        elif status == "tracing-stop":
            return cls.TracingStop
        elif status == "zombie":
            return cls.Zombie
        elif status == "dead":
            return cls.Dead
        elif status == "wake-kill":
            return cls.WakeKill
        elif status == "waking":
            return cls.Waking
        elif status == "idle":  # Linux, macOS, FreeBSD
            return cls.Idle
        elif status == "locked":  # FreeBSD
            return cls.Locked
        elif status == "waiting":  # FreeBSD
            return cls.Waiting
        elif status == "suspended":  # NetBSD
            return cls.Suspended
        elif status == "parked":  # Linux
            return cls.Parked
        else:
            return cls.Unknown


@dataclass
class DaemonA:
    plugin: str
    slug: str
    name: str
    address: str
    description: str
    enable: bool
    created_at: datetime
    updated_at: datetime

    state: DaemonState
    exit_code: Optional[int] = None


@dataclass
class CreateDaemonQ:
    plugin: str
    slug: str
    name: Optional[str] = None
    address: Optional[str] = None
    description: Optional[str] = None
    enable: bool = False


@dataclass
class UpdateDaemonQ:
    slug: Optional[str] = None
    name: Optional[str] = None
    address: Optional[str] = None
    description: Optional[str] = None
    enable: Optional[bool] = False

File: api/src/packet/daemon.ts
export enum DaemonState {
  Unknown = 0,
  Unregistered = 1,
  Down = 2,
  NoSuchProcess = 3,

  Running = 101,
  Sleeping = 102,
  DiskSleep = 103,
  Stopped = 104,
  TracingStop = 105,
  Zombie = 106,
  Dead = 107,
  WakeKill = 108,
  Waking = 109,
  Idle = 110,
  Locked = 111,
  Waiting = 112,
  Suspended = 113,
  Parked = 114,
}

export function getStateName(state: DaemonState) {
  switch (state) {
    case DaemonState.Unknown:
      return 'Unknown';
    case DaemonState.Unregistered:
      return 'Unregistered';
    case DaemonState.Down:
      return 'Down';
    case DaemonState.NoSuchProcess:
      return 'NoSuchProcess';
    case DaemonState.Running:
      return 'Running';
    case DaemonState.Sleeping:
      return 'Sleeping';
    case DaemonState.DiskSleep:
      return 'DiskSleep';
    case DaemonState.Stopped:
      return 'Stopped';
    case DaemonState.TracingStop:
      return 'TracingStop';
    case DaemonState.Zombie:
      return 'Zombie';
    case DaemonState.Dead:
      return 'Dead';
    case DaemonState.WakeKill:
      return 'WakeKill';
    case DaemonState.Waking:
      return 'Waking';
    case DaemonState.Idle:
      return 'Idle';
    case DaemonState.Locked:
      return 'Locked';
    case DaemonState.Waiting:
      return 'Waiting';
    case DaemonState.Suspended:
      return 'Suspended';
    case DaemonState.Parked:
      return 'Parked';
    default:
      return `UnknownState(${state})`;
  }
}

export function getStateColor(state: DaemonState) {
  switch (state) {
    case DaemonState.Unknown:
      return 'red';
    case DaemonState.Unregistered:
    case DaemonState.Down:
    case DaemonState.NoSuchProcess:
      return 'grey';
    case DaemonState.Running:
      return 'green';
    case DaemonState.Sleeping:
    case DaemonState.DiskSleep:
    case DaemonState.Stopped:
    case DaemonState.TracingStop:
      return 'blue';
    case DaemonState.Zombie:
      return 'deep-orange';
    case DaemonState.Dead:
    case DaemonState.WakeKill:
    case DaemonState.Waking:
    case DaemonState.Idle:
    case DaemonState.Locked:
    case DaemonState.Waiting:
    case DaemonState.Suspended:
    case DaemonState.Parked:
      return 'grey';
    default:
      return 'red';
  }
}

export function isStateRunning(state: DaemonState) {
  switch (state) {
    case DaemonState.Unknown:
    case DaemonState.Unregistered:
    case DaemonState.Down:
    case DaemonState.NoSuchProcess:
      return false;
    case DaemonState.Running:
    case DaemonState.Sleeping:
    case DaemonState.DiskSleep:
    case DaemonState.Stopped:
    case DaemonState.TracingStop:
      return true;
    case DaemonState.Zombie:
      return false;
    case DaemonState.Dead:
      return false;
    case DaemonState.WakeKill:
    case DaemonState.Waking:
    case DaemonState.Idle:
    case DaemonState.Locked:
    case DaemonState.Waiting:
    case DaemonState.Suspended:
    case DaemonState.Parked:
      return true;
    default:
      return false;
  }
}

export interface DaemonA {
  plugin: string;
  slug: string;
  name: string;
  address: string;
  description: string;
  enable: boolean;
  created_at: string;
  updated_at: string;

  state: DaemonState;
  exit_code?: number;
}

export interface CreateDaemonQ {
  plugin: string;
  slug: string;
  name?: string;
  address?: string;
  description?: string;
  enable: boolean;
}

export interface UpdateDaemonQ {
  slug?: string;
  name?: string;
  address?: string;
  description?: string;
  enable?: boolean;
}
